src/transport/ExchangeMgr.h

```diff
@@ -0,0 +1,367 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the classes corresponding to CHIP Exchange management.
+ *
+ */
+
+#ifndef EXCHANGE_MGR_H
+#define EXCHANGE_MGR_H
+
+#include <support/DLLUtil.h>
+#include <system/SystemTimer.h>
+#include <transport/SecureSessionMgr.h>
+
+#define EXCHANGE_CONTEXT_ID(x) ((x) + 1)
+
+using namespace chip::Transport;
+
+namespace chip {
+
+using System::PacketBuffer;
+
+class ExchangeManager;
+
+/**
+ *  @class ChipExchangeHeader
+ *
+ *  @brief
+ *    This specifies the CHIP profile and message type of a particular CHIP message
+ *    within an ExchangeContext.
+ */
+class ChipExchangeHeader
+{
+public:
+    uint8_t Version;     /**< The version of the CHIP Exchange Header format */
+    uint8_t Flags;       /**< Bit flag indicators for the type of CHIP message */
+    uint16_t ExchangeId; /**< The Exchange identifier for the ExchangeContext */
+    uint32_t ProfileId;  /**< The Profile identifier of the CHIP message */
+    uint8_t MessageType; /**< The Message type for the specified CHIP profile */
+};
+
+/**
+ *  @brief
+ *    The CHIP Exchange header version.
+ */
+typedef enum ChipExchangeVersion
+{
+    kChipExchangeVersion_V1 = 1
+} ChipExchangeVersion;
+
+/**
+ *  @brief
+ *    The CHIP Exchange header flag bits.
+ */
+typedef enum ChipExchangeFlags
+{
+    kChipExchangeFlag_Initiator = 0x1, /**< Set when current message is sent by the initiator of an exchange */
+    kChipExchangeFlag_AckId     = 0x2, /**< Set when current message is an acknowledgment for a previously received message */
+    kChipExchangeFlag_NeedsAck  = 0x4  /**< Set when current message is requesting an acknowledgment from the recipient. */
+} ChipExchangeFlags;
+
+/**
+ *  @class ExchangeContext
+ *
+ *  @brief
+ *    This class represents an ongoing conversation (ExchangeContext) between two or more nodes.
+ *    It defines methods for encoding and communicating CHIP messages within an ExchangeContext
+ *    over various transport mechanisms, for example, TCP, UDP, or CHIP Reliable Messaging.
+ *
+ */
+class DLL_EXPORT ExchangeContext
+{
+    friend class ExchangeManager;
+
+public:
+    typedef uint32_t Timeout; /**< Type used to express the timeout in this ExchangeContext, in milliseconds */
+
+    ExchangeManager * ExchangeMgr; /**< [READ ONLY] Owning exchange manager. */
```

Comment:
Drive by review - I realize this is still in draft.

Please no public data members. Also limit data members (e.g. CRMP logic should be handled by separate classes).

Reply:
ExchangeContext is semantically used by ExchangeManager and protocols as a struct more than a class. I have cleanup the CRMP (wrapped by CHIP_CONFIG_ENABLE_RELIABLE_MESSAGING flag) from there interface.

Reply:
I believe my original request still stands - there should be no public data members at all. We can create inline accessors if needed. Public data members are very hard to reason about.

Reply:
Acked